### PR TITLE
Separate DRPolicy into DRCluster and a policy

### PR DESCRIPTION
**Pull request by ShyamsundarR**
This PR is to separate DRPolicy into 2 CRDs, a DRPolicy CRD and a per managed cluster DRCluster CRD.

A DRCluster has a 1:1 relationship with a ManagedCluster and carries cluster level specification. Further it also serves to collect and represent cluster level storage data (e.g storage IDs).

A DRPolicy can have more than one DRCluster, and multiple DRPolicies can refer to the same DRCluster as well. A policy only governs how data is to be protected and on which clusters.

Further issues with the older DRPolicy containing cluster data were the n:1 DRPolicy:ManagedCluster relationship. For example a cluster fenced in a DRPolicy can remain unfenced in another.

This patch series hence splits the DRCluster related specification into its own resource, and retains only policy related specification in DRPolicy.

NOTE: It does not aim to change anything else at present and subsequent PRs would leverage this scheme for future work.

TODO:

- [x] Fix envtests (currently fail due to the refactor as expected)
- [x] Run e2e and ensure components work as expected
- [x] Remove DRPolicy.Spec.DRClusterSet

Future PRs:
- [ ] Think through correct condition names for DRCluster and if required DRPolicy
- [ ] Address any pending TODOs
- [ ] Fetch StorageClass and VolumeReplicationClass storage ID labels from managed clusters and update DRCluster status with the same
- [ ] Amend sync/async determination in code based on storage IDs fetched
- [ ] Pass labels to VRG to ensure PVCs adhere to supported CSI drivers
- [ ] Remove DRCluster.Spec.Region as a result of the above

### CODE REVIEW

File: PROJECT [hunk 1]
```diff
@@ -34,4 +34,13 @@ resources:
   kind: DRPlacementControl
   path: github.com/ramendr/ramen/api/v1alpha1
   version: v1alpha1
+- api:
+    crdVersion: v1
+    namespaced: true
```
Comment (raghavendra-talur) [inline PROJECT:39]: We think DRCluster should be scoped similar to DRPolicy and also because it is a representation of a managed cluster, it should not be namespaced.

File: controllers/drclusters.go [hunk 2]
```diff
@@ -120,7 +120,7 @@ func objectsToDeploy(hubOperatorRamenConfig *rmn.RamenConfig) ([]interface{}, er
 	ramenConfig := &drClusterOperatorRamenConfig
 	drClusterOperatorNamespaceName := drClusterOperatorNamespaceNameOrDefault(ramenConfig)
 	ramenConfig.LeaderElection.ResourceName = drClusterLeaderElectionResourceName
-	ramenConfig.RamenControllerType = rmn.DRCluster
+	ramenConfig.RamenControllerType = rmn.DRClusterType
```
Comment (raghavendra-talur) [inline controllers/drclusters.go:123]: In addition to renaming this variable, it would also be good to rename this file to avoid confusion.

File: api/v1alpha1/drpolicy_types.go [hunk 3]
```diff
@@ -81,7 +81,10 @@ type DRPolicySpec struct {
 
 	// The set of managed clusters governed by this policy, which have
 	// replication relationship enabled between them.
-	DRClusterSet []ManagedCluster `json:"drClusterSet"`
+	DRClusterSet []ManagedCluster `json:"drClusterSet,omitempty"`
```
Comment (raghavendra-talur) [inline api/v1alpha1/drpolicy_types.go:84]: Should this line be deleted?

File: controllers/drclusters.go (hunk 2 above)
Comment (ShyamsundarR) [reply controllers/drclusters.go:123]: later once refactor is complete,  but yes possibly split into 2 files drpolicy and drcluster.

File: api/v1alpha1/drpolicy_types.go (hunk 3 above)
Comment (ShyamsundarR) [reply api/v1alpha1/drpolicy_types.go:84]: Yes, but need to finish the refactor first, for example the second commit does that. This can now be deleted, will add to the task list, but retained in the first pass as otherwise code wont compile.

File: PROJECT (hunk 1 above)
Comment (ShyamsundarR) [reply PROJECT:39]: I had/have reservations on the scope, and wanted to be sure we want this at cluster scope and not namespaced:
- Do we need users of the cluster to be able to see the resource?
- Can a cluster scoped resource (DRPolicy) point to a namespaced resource (DRCluster)?

I have initially chosen a namespace scope, to see what problems we may encounter during implementation (do not see any yet other than when we attempt to list all DRClusters.

Anyway long story short, I did not want to apply ManagedCluster scope rules to this one, but wanted more reasons for its scope to be at the cluster level, and I am still debating what is correct.

File: api/v1alpha1/drcluster_types.go [hunk 4]
```diff
@@ -48,6 +48,7 @@ type DRClusterSpec struct {
 // DRClusterStatus defines the observed state of DRCluster
 type DRClusterStatus struct {
 	Conditions []metav1.Condition `json:"conditions,omitempty"`
+	Fenced     FenceStatus        `json:"status,omitempty"`
```
Comment (ShyamsundarR) [inline api/v1alpha1/drcluster_types.go:51]: NOTE: Changing this to a condition is not to be part of this PR

File: controllers/drpolicy_controller.go [hunk 5]
```diff
@@ -213,221 +200,76 @@ func hasConflictingDRPolicy(match *ramen.DRPolicy, list ramen.DRPolicyList) erro
...
-
-	if len(invalidCidrs) > 0 {
-		return fmt.Errorf("cluster %s has invalid CIDRs %s", cluster.Name, strings.Join(invalidCidrs, ", "))
-	}
-
-	return nil
-}
-
-type objectUpdater struct {
+type drpolicyUpdater struct {
```
Comment (ShyamsundarR) [inline controllers/drpolicy_controller.go:228]: NOTE: The type is made Kind specific as accessing <kind>.Status is not generic and depends on the type. It is conceivable that we can write functions in the base type and create a new interface and hence retain a single type here, but the interfaces would get exposed in the same package as v1alpha1 for ramen, hence did not do that.

Changing it so and adding some generic routines seemed the optimal middle path for this refactor.

File: controllers/drclusters.go (hunk 2 above)
Comment (ShyamsundarR) [reply controllers/drclusters.go]: Done

File: api/v1alpha1/drpolicy_types.go (hunk 3 above)
Comment (ShyamsundarR) [reply api/v1alpha1/drpolicy_types.go]: Done

**Comment by ShyamsundarR** [comment]
@raghavendra-talur @raghavendrabhat Completed e2e run with relocate-failover-relocate, including initial cluster deployment automation and secrets distribution automation tasks. Had a minor change to the e2e scripts due to the same which is now amended in the last commit. Ready for the final review and merge as required.

**Comment by ShyamsundarR** [comment]
Related samples PR is here: https://github.com/RamenDR/ocm-ramen-samples/pull/14

**Review by BenamarMk — APPROVED** [review]
